### PR TITLE
[MRG] Adding BIOT to braindecode

### PR DESCRIPTION
**Pull request by bruAristimunha**

### CODE REVIEW

**Comment by bruAristimunha** [comment]
FYI @JessyD

File: braindecode/models/biot.py [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+            # The positional embedding is explain with more
+            # detail in the _PositionalEncoding class.
+            channel_emb = self.positional_encoding(
+                channel_spec_emb + channel_token_emb)
+            # In case of perturb, the time steps are randomly selected
+            # and the channel embedding is reduced to a random number
+            # of time steps.
+            if perturb:
+                ts = channel_emb.shape[1]
+                ts_new = np.random.randint(ts // 2, ts)
```
Comment (agramfort) [inline braindecode/models/biot.py:297]: can you rather use torch here?

**Review by agramfort — COMMENTED** [review]
you'll need to add tests. I would also ping on the PR the original authors if they want to have a look

File: braindecode/models/biot.py [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+        The number of Fourier transform points
+    hop_length: int (default 100)
+        The distance between neighboring sliding window frames
+    """
+
+    def __init__(
+            self,
+            emb_size=256, # The size of the embedding layer
+            att_num_heads=8, # The number of attention heads
+            n_chans=16, # The number of channels
```
[the remark this answers is not in the document]
Comment (bruAristimunha) [reply braindecode/models/biot.py:183]: I change the order

File: test/unit_tests/models/test_models.py [hunk 3]
```diff
@@ -475,7 +479,7 @@ def test_eldele_2021(sfreq, n_classes, input_size_s, d_model):
     # (100, 5, 30, 80) - Physionet Sleep
     # (125, 4, 30, 100) - SHHS
     rng = np.random.RandomState(42)
-    n_channels = 1
+    n_channels = 10
```
Comment (bruAristimunha) [inline test/unit_tests/models/test_models.py:482]:
```suggestion
    n_channels = 1
```

File: test/integration_tests/test_models_correctness.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    preload=True,
+    )
+
+    # Convert data from volt to millivolt
+    # Pytorch expects float32 for input and int64 for labels.
+    X = (epoched.get_data() * 1e6).astype(np.float32)
+    y = (epoched.events[:, 2] - 2).astype(np.int64)  # 2,3 -> 0,1
+    return X, y
+
+def test_correctness_biot(real_data):
```
Comment (bruAristimunha) [inline test/integration_tests/test_models_correctness.py:70]: I think you will like this test @tgy

**Comment by bruAristimunha** [comment]
pinging in can you want to give any feedback @ycq091044

**Comment by bruAristimunha** [comment]
@agramfort, I am happy with the PR; it is okay for me to merge. Can you take a last look? 🙏🏽

File: braindecode/models/biot.py [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+            warn("This model has only been trained on a dataset with 200 Hz. " +
+                 "no guarantee to generalize well with the default parameters",
+                 UserWarning)
+        if self.n_chans > emb_size:
+            warn("The number of channels is larger than the embedding size. " +
+                 "This may cause overfitting. Consider using a larger " +
+                 "embedding size or a smaller number of channels.",
+                 UserWarning)
+
+        self.biot = _BIOTEncoder(emb_size=emb_size,
```
Comment (agramfort) [inline braindecode/models/biot.py:408]:
```suggestion
        self.encoder = _BIOTEncoder(emb_size=emb_size,
```

File: braindecode/models/biot.py [hunk 6]
```diff
@@ -0,0 +1,419 @@
...
+                                 att_num_heads=att_num_heads,
+                                 n_layers=n_layers, n_chans=self.n_chans,
+                                 n_fft=self.sfreq, hop_length=hop_length)
+
+        self.classifier = _ClassificationHead(emb_size=emb_size,
+                                              n_outputs=self.n_outputs)
+
+    def forward(self, x):
```
Comment (agramfort) [inline braindecode/models/biot.py:416]: docstring?

File: test/integration_tests/test_models_correctness.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
+# Authors: Bruno Aristimunha
```
Comment (agramfort) [inline test/integration_tests/test_models_correctness.py:1]: can you run black on this file? it violates a number of pep8 rules

File: test/integration_tests/test_models_correctness.py (hunk 4 above)
Comment (agramfort) [reply test/integration_tests/test_models_correctness.py:70]: how long does it take to run? it's not too slow for CIs?

File: test/integration_tests/test_models_correctness.py (hunk 7 above)
Comment (agramfort) [reply test/integration_tests/test_models_correctness.py:1]: also why a few folder in test? as you only test biot I would call this file test_biot.py 

i would have kept it in unit_tests/models/

Comment (bruAristimunha) [reply test/integration_tests/test_models_correctness.py:1]: I was thinking of creating more tests about full integration in the future, but you are right. It is a topic for another PR.

File: test/integration_tests/test_models_correctness.py (hunk 4 above)
Comment (bruAristimunha) [reply test/integration_tests/test_models_correctness.py:70]: 12 seconds.

The dataset generation is taking longer from python 3.8:

19.85s setup    test/integration_tests/test_models_correctness.py::test_correctness_biot
12.11s call     test/integration_tests/test_models_correctness.py::test_correctness_biot

Comment (bruAristimunha) [reply test/integration_tests/test_models_correctness.py]: At the moment, I think we are fine with time in the test and doc, I spent some time optimizing the CI during the last sprint.

**Comment by bruAristimunha** [comment]
The green bottom is your @agramfort =)

**Review by agramfort — APPROVED** [review]
cc @ycq091044

**Comment by agramfort** [comment]
thx @bruAristimunha

**Pull request merged**